[docs] Tune overhaul part II (#22656)
ray-project/ray: doc/source/tune/key-concepts.rst

Key Concepts
============

.. TODO: should we introduce checkpoints as well?
.. TODO: should we at least mention "Stopper" classes here?

Let's quickly walk through the key concepts you need to know to use Tune.
If you want to see practical tutorials right away, go visit our :ref:`user guides<tune-guides>`.
In essence, Tune has six crucial components that you need to understand.

First, you define the hyperparameters you want to tune in a `search space` and pass them into a ``Trainable``
that specifies the objective you want to tune.
Then you select a `search algorithm` to effectively optimize your parameters and optionally use a
`scheduler` to stop searches early and speed up your experiments.
Together with other configuration, your ``Trainable``, algorithm, and scheduler are passed into ``tune.run()``,
which runs your experiments and creates `trials`.

> These trials can then be used in `analyses` to inspect your experiment results.

Reviewer: "Trainable" should have the same formatting as the names of the other components IMO

Author: @Yard1  The single back ticks emphasise, double back ticks stand for code. Trainable is a class, hence this convention. Then again, there's a certain ambiguity because we use "trainable" very colloquially here as well, e.g. for the function API.

In a way, I feel like "trainable" is a relatively poor convention. We'd generally be better off to call this concept simply an "objective".

Anyway, not sure what's right here.

Reviewer: Yeah, I am aware of the distinction between single and double backtics - my issue here is that Trainable is the only name that is being formatted as code, despite the possibility of formatting the other names as well. We should either format everything as classes, or format nothing IMO.

Author: understood. addressed that point 👍

The following figure shows an overview of these components, which we cover in detail in the next sections.

.. image:: images/tune_flow.png

Trainables
----------

In short, a :ref:`Trainable<trainable-docs>` is an object that you can pass into a Tune run.
Ray Tune has two ways of defining a ``Trainable``, namely the :ref:`Function API <tune-function-api>`
and the :ref:`Class API<tune-class-api>`.

> Both are valid ways of defining a ``Trainable``, but the ``Function API`` is generally recommended and is used

Reviewer: ```suggestion
Both are valid ways of defining a ``Trainable``, but the *Function API* is generally recommended and is used
```

        :language: python
        :start-after: __class_api_start__
        :end-before: __class_api_end__

    .. tip:: ``tune.report`` can't be used within a ``Trainable`` class.

Learn more about the details of :ref:`Trainables here<trainable-docs>`
and :ref:`have a look at our examples <tune-general-examples>`.
Next, let's have a closer look at what the ``config`` dictionary is that you pass into your trainables.

Search Spaces
-------------

To optimize your *hyperparameters*, you have to define a *search space*.
A search space defines valid values for your hyperparameters and can specify
how these values are sampled (e.g. from a uniform distribution or a normal
distribution).

Tune offers various functions to define search spaces and sampling methods.
:ref:`You can find the documentation of these search space definitions here <tune-sample-docs>`.

Here's an example covering all search space functions. Again,
:ref:`here is the full explanation of all these functions <tune-sample-docs>`.

.. literalinclude:: doc_code/key_concepts.py
    :language: python
    :start-after: __config_start__
    :end-before: __config_end__

Trials
------

> You use :ref:`tune.run <tune-run-ref>` to execute and manage hyperparameter tuning and generate your `trials`.

Reviewer: can we make ":ref:\`tune.run \<tune-run-ref\>\`" have the same formatting as \`\`tune.run()\`\`?

Author: I believe it's not possible, but let me check.

Author: not possible :/

At a minimum, your ``tune.run()`` call takes in a trainable as first argument, and a ``config`` dictionary
to define your search space.

The ``tune.run()`` function also provides many features such as :ref:`logging <tune-logging>`,
:ref:`checkpointing <tune-checkpoint-syncing>`, and :ref:`early stopping <tune-stopping-ref>`.
Continuing with the example defined earlier (minimizing ``a (x ** 2) + b``), a simple Tune run with a simplistic
search space for ``a`` and ``b`` would look like this:

.. literalinclude:: doc_code/key_concepts.py
    :language: python
    :start-after: __run_tunable_start__
    :end-before: __run_tunable_end__

``tune.run`` will generate a couple of hyperparameter configurations from its arguments,
wrapping them into :ref:`Trial objects <trial-docstring>`.

Trials contain a lot of information.
For instance, you can get the hyperparameter configuration used (``trial.config``), the trial ID (``trial.trial_id``),
the trial's resource specification (``resources_per_trial`` or ``trial.placement_group_factory``) and many other values.

> ``tune.run`` will execute until all trials stop or error.

Reviewer: ```suggestion
By default ``tune.run`` will execute until all trials stop or error.
```

    +----------------------+----------+---------------------+-----------+--------+--------+----------------+-------+
    | Trial name           | status   | loc                 |         a |      b |  score | total time (s) |  iter |
    |----------------------+----------+---------------------+-----------+--------+--------+----------------+-------|
    | Trainable_a826033a | RUNNING  | 10.234.98.164:31115 | 0.303706  | 0.0761 | 0.1289 |        7.54952 |    15 |
    +----------------------+----------+---------------------+-----------+--------+--------+----------------+-------+

> You can also easily run just 10 trials by specifying the number of samples (``num_samples``).

Reviewer: Should we mention `time_budget_s` here as an alternative? (with `num_samples=-1`) A lot of people don't know about that

Author: I didn't know that :D 

Maybe as a side note? I don't want to overload the key concepts section.

Reviewer: I think it's important enough that it should be front and center.

Author: I added a sentence

Tune automatically :ref:`determines how many trials will run in parallel <tune-parallelism>`.

.. literalinclude:: doc_code/key_concepts.py
    :language: python
    :start-after: __run_tunable_samples_start__
    :end-before: __run_tunable_samples_end__

Finally, you can use more interesting search spaces to optimize your hyperparameters
via Tune's :ref:`search space API <tune-default-search-space>`, like using random samples or grid search.
Here's an example of uniformly sampling between ``[0, 1]`` for ``a`` and ``b``:

.. literalinclude:: doc_code/key_concepts.py
    :language: python
    :start-after: __search_space_start__
    :end-before: __search_space_end__

To learn more about the various ways of configuring your Tune runs,
check out the :ref:`tune.run() API reference<tune-run-ref>`.

Search Algorithms
-----------------

To optimize the hyperparameters of your training process, you use
a :ref:`Search Algorithm <tune-search-alg>` which suggests hyperparameter configurations.
If you don't specify a search algorithm, Tune will use random search by default, which can provide you
with a good starting point for your hyperparameter optimization.

> For instance, to use Tune with Bayesian optimization (make sure to first run ``pip install bayesian-optimization``),

Reviewer: Technically all of the searchers we have are bayesian optimization. Maybe add "simple" or specify the name of library

we can define an ``algo`` using ``BayesOptSearch``.
Simply pass in a ``search_alg`` argument to ``tune.run``:

.. literalinclude:: doc_code/key_concepts.py
    :language: python
    :start-after: __bayes_start__
    :end-before: __bayes_end__

> Tune has SearchAlgorithms that integrate with many popular **optimization** libraries,

Reviewer: ```suggestion
Tune has Search Algorithms that integrate with many popular **optimization** libraries,
```

> such as :ref:`Nevergrad <nevergrad>` and :ref:`HyperOpt <tune-hyperopt>`.

Reviewer: Definitely add Optuna here

Tune automatically converts the provided search space into the search

> spaces the search algorithms and underlying libraries expect.

Reviewer: Maybe we should mention that some samplers are not supported, eg. the `sample_from` one

Author: not here

   * - :ref:`TuneBOHB <suggest-TuneBOHB>`
     - Bayesian Opt/HyperBand
     - [`BOHB <https://github.com/automl/HpBandSter>`__]
     - :doc:`/tune/examples/includes/bohb_example`
   * - :ref:`NevergradSearch <nevergrad>`
     - Gradient-free Optimization
     - [`Nevergrad <https://github.com/facebookresearch/nevergrad>`__]
     - :doc:`/tune/examples/includes/nevergrad_example`
   * - :ref:`OptunaSearch <tune-optuna>`
     - Optuna search algorithms
     - [`Optuna <https://optuna.org/>`__]
     - :doc:`/tune/examples/includes/optuna_example`
   * - :ref:`ZOOptSearch <zoopt>`
     - Zeroth-order Optimization
     - [`ZOOpt <https://github.com/polixir/ZOOpt>`__]
     - :doc:`/tune/examples/includes/zoopt_example`
   * - :ref:`SigOptSearch <sigopt>`
     - Closed source
     - [`SigOpt <https://sigopt.com/>`__]
     - :doc:`/tune/examples/includes/sigopt_example`
   * - :ref:`HEBOSearch <tune-hebo>`
     - Heteroscedastic Evolutionary Bayesian Optimization
     - [`HEBO <https://github.com/huawei-noah/HEBO/tree/master/HEBO>`__]
     - :doc:`/tune/examples/includes/hebo_example`

.. note:: Unlike :ref:`Tune's Trial Schedulers <tune-schedulers>`,

>     Tune SearchAlgorithms cannot affect or stop training processes.

Reviewer: ```suggestion
    Tune Search Algorithms cannot affect or stop training processes.
```

you can :ref:`read the instructions here <byo-algo>`.

Tune also provides helpful utilities to use with Search Algorithms:

 * :ref:`repeater`: Support for running each *sampled hyperparameter* with multiple random seeds.
 * :ref:`limiter`: Limits the amount of concurrent trials when running optimization.
 * :ref:`shim`: Allows creation of the search algorithm object given a string.

Note that in the example above we  tell Tune to ``stop`` after ``20`` training iterations.
This way of stopping trials with explicit rules is useful, but in many cases we can do even better with
`schedulers`.

.. _schedulers-ref:

Schedulers
----------

To make your training process more efficient, you can use a :ref:`Trial Scheduler <tune-schedulers>`.
For instance, in our ``trainable`` example minimizing a function in a training loop, we used ``tune.report()``.
This reported `incremental` results, given a hyperparameter configuration selected by a search algorithm.
Based on these reported results, a Tune scheduler can decide whether to stop the trial early or not.
If you don't specify a scheduler, Tune will use a first-in-first-out (FIFO) scheduler by default, which simply

> passes through the trials selected by your search algorithm in the order they were picked.

Reviewer: ```suggestion
passes through the trials selected by your search algorithm in the order they were picked and does not perform any early stopping.
```

To use a scheduler, just pass in a ``scheduler`` argument to ``tune.run()``:

.. literalinclude:: doc_code/key_concepts.py
    :language: python
    :start-after: __hyperband_start__
    :end-before: __hyperband_end__

Tune includes distributed implementations of early stopping algorithms such as
`Median Stopping Rule <https://research.google.com/pubs/pub46180.html>`__, `HyperBand <https://arxiv.org/abs/1603.06560>`__,
and `ASHA <https://openreview.net/forum?id=S1Y7OOlRZ>`__.
Tune also includes a distributed implementation of `Population Based Training (PBT) <https://deepmind.com/blog/population-based-training-neural-networks>`__
and `Population Based Bandits (PB2) <https://arxiv.org/abs/2002.02518>`__.

.. tip:: The easiest scheduler to start with is the ``ASHAScheduler`` which will aggressively terminate low-performing trials.

When using schedulers, you may face compatibility issues, as shown in the below compatibility matrix.
Certain schedulers cannot be used with search algorithms,
and certain schedulers require :ref:`checkpointing to be implemented <tune-checkpoint-syncing>`.

Schedulers can dynamically change trial resource requirements during tuning.

> This is currently implemented in ``ResourceChangingScheduler``, which can wrap around any other scheduler.

Reviewer: can we ref ResourceChangingScheduler?
